Split Dynamic Resource Allocation into a Resource Management section (#56295)
kubernetes/website: content/en/docs/concepts/resource-management/dynamic-resource-allocation/_index.md

---
reviewers:
- klueska
- pohly
title: Dynamic Resource Allocation
content_type: concept
weight: 20
aliases:
- /docs/concepts/scheduling-eviction/dynamic-resource-allocation/
---

<!-- overview -->

{{< feature-state feature_gate_name="DynamicResourceAllocation" >}}

This section introduces _dynamic resource allocation (DRA)_ in Kubernetes.

{{< glossary_definition prepend="DRA is" term_id="dra" length="all" >}}

Allocating resources with DRA offers a similar experience to
[dynamic volume provisioning](/docs/concepts/storage/dynamic-provisioning/),
in which you use PersistentVolumeClaims to _claim_ storage capacity from storage classes,
and request the claimed capacity for use in your Pods.

<!-- body -->

### Benefits of DRA {#dra-benefits}

DRA provides a flexible way to categorize, request, and use devices in your cluster.
Using DRA provides benefits like the following:

* **Flexible device filtering**: use common expression language (CEL) to perform
  fine-grained filtering for specific device attributes.
* **Device sharing**: share the same resource with multiple containers or Pods

>   by referencing the corresponding resource claim.

Reviewer: ```suggestion
  by referencing the corresponding resource claim.
* **Device configuration**: attach vendor-specific device configurations to your resource claim, enabling per-workload device configuration, rather than today's per-node device configuration
```

  device class for general-purpose workloads, and a high-performance device
  class for critical jobs.
* **Simplified Pod requests**: with DRA, app operators don't need to specify
  device quantities in Pod resource requests. Instead, the Pod references a
  resource claim, and the device configuration in that claim applies to the Pod.

These benefits provide significant improvements in the device allocation
workflow when compared to
[device plugins](/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/),
which require per-container device requests, don't support device sharing, and
don't support expression-based device filtering.

### Types of DRA users {#dra-user-types}

The workflow of using DRA to allocate devices involves the following types of users:

* **Device owner**: responsible for devices. Device owners might be commercial
  vendors, the cluster operator, or another entity. To use DRA, devices must
  have DRA-compatible drivers that do the following:

  * Create ResourceSlices that provide Kubernetes with information about
    nodes and resources.
  * Update ResourceSlices when resource capacity in the cluster changes.

>   * Optionally, create DeviceClasses that workload operators can use to

Reviewer: ```suggestion
  * Configure devices according to the claim, and attach them to containers via Container Device Interface (CDI).
  * Optionally, create DeviceClasses that workload operators can use to
```

---
reviewers:
- klueska
- pohly
title: Dynamic Resource Allocation
content_type: concept
weight: 20
aliases:
- /docs/concepts/scheduling-eviction/dynamic-resource-allocation/
---

<!-- overview -->

{{< feature-state feature_gate_name="DynamicResourceAllocation" >}}

This section introduces _dynamic resource allocation (DRA)_ in Kubernetes.

> <!-- body -->
> 
> ## About DRA {#about-dra}
> 
> {{< glossary_definition prepend="DRA is" term_id="dra" length="all" >}}
> 
> Allocating resources with DRA is a similar experience to
> [dynamic volume provisioning](/docs/concepts/storage/dynamic-provisioning/),
> in which you use PersistentVolumeClaims to claim storage capacity from storage classes
> and request the claimed capacity in your Pods.
> 

Reviewer 2: One early piece of feedback
```suggestion
This section introduces _dynamic resource allocation (DRA)_ in Kubernetes.

{{< glossary_definition prepend="DRA is" term_id="dra" length="all" >}}

Allocating resources with DRA offers a similar experience to
[dynamic volume provisioning](/docs/concepts/storage/dynamic-provisioning/),
in which you use PersistentVolumeClaims to _claim_ storage capacity from storage classes,
and request the claimed capacity for use in your Pods.

<!-- body -->

```

_It can wait_. Just highlighting that the suggestion is closer to how I see our desired state.
